### DISCUSSION TITLE
Relation between mat.elastic_tensor.ieee_format and mat.elastic_tensor.raw

### DISCUSSION THREAD

**Discussion by fxcoudert** [materialsproject/pymatgen#4058] [General]
I have two questions related to the elastic tensors representation in Materials Project/pymatgen.

1. Why is `mat.elastic_tensor.ieee_format` rounded to 1 GPa for all entries?
2. Where can I find the exact relationship between `mat.elastic_tensor.ieee_format` and `mat.elastic_tensor.raw`? What function is transforming one into the other? I would like to do it on my own, without rounding.

**Comment by janosh** [comment]
maybe @matthewkuner knows more?

**Comment by matthewkuner** [comment]
1)
I do not know why the values are rounded on the website; honestly, it is probably just to make them more pretty. **However**, the data from the mp-api is _not_ rounded. Try 
```
with MPRester("insert_your_mp_api_key_here") as mpr:
    elastic_doc = mpr.materials.elasticity.search(material_ids = ["mp-149"])
print(elastic_doc.elastic_tensor)
```

2)
I **_think_** that the ieee format symmetrizes the elastic tensor based on the symmetry of the structure. Coding this yourself will be a decent bit of work--I would not recommend that at all, as it has certainly been done countless times in open-source repos already (e.g. [here](https://github.com/materialsproject/pymatgen/blob/1ff1ba594960c2cc3c60839780d47ff8392ee65c/src/pymatgen/core/tensors.py#L478) in pymatgen ). If you want to understand this better, the root of the relationship between those tensors would be grounded in group/representation theory, crystal symmetry, tensor transformations, etc.

Somebody please correct me if I'm wrong. I work with this stuff somewhat but am by no means an expert...

**Reply by hongyi-zhao** [reply]
> ```python
> with MPRester("insert_your_mp_api_key_here") as mpr:
>     elastic_doc = mpr.materials.elasticity.search(material_ids = ["mp-149"])
> print(elastic_doc.elastic_tensor)
> ```

The `print(elastic_doc.elastic_tensor)` line should be written as follows:

```python
In [5]: from mp_api.client import MPRester
   ...: 
   ...: with MPRester() as mpr:
   ...:     elastic_doc = mpr.materials.elasticity.search(material_ids=["mp-149"
   ...: ])
   ...:     # Since search returns a list, we need to access the first element
   ...:     if elastic_doc:  # Check if the list is not empty
   ...:         print(elastic_doc[0].elastic_tensor)  # Access the first documen
   ...: t's elastic tensor
   ...:     else:
   ...:         print("No elastic data found for mp-149")
   ...: 
Retrieving ElasticityDoc documents: 100%|███████| 1/1 [00:00<00:00, 9686.61it/s]
raw=((153.19279267761982, 56.777185976088205, 56.777185976088205, -7.979727989493311e-17, -6.661338147750937e-16, 9.436895709313829e-16), (56.777185976088205, 153.19279267761982, 56.77718597608822, -6.661338147750937e-16, -7.979727989493311e-17, 6.839865079851815e-15), (56.777185976088205, 56.77718597608822, 153.19279267761985, -6.661338147750941e-16, 9.436895709313833e-16, -7.979727989493315e-17), (-7.979727989493311e-17, -6.661338147750937e-16, -6.661338147750941e-16, 74.25053622362908, 0.0, 1.2528910738651292e-32), (-6.661338147750937e-16, -7.979727989493311e-17, 9.436895709313833e-16, 0.0, 74.2505362236291, -2.0608663726559307e-15), (9.436895709313829e-16, 6.839865079851815e-15, -7.979727989493315e-17, 1.2528910738651292e-32, -2.0608663726559307e-15, 74.25053622362913)) ieee_format=((153.0, 57.0, 57.0, -0.0, -0.0, 0.0), (57.0, 153.0, 57.0, -0.0, -0.0, 0.0), (57.0, 57.0, 153.0, -0.0, 0.0, 0.0), (-0.0, -0.0, -0.0, 74.0, 0.0, -0.0), (-0.0, -0.0, 0.0, 0.0, 74.0, -0.0), (0.0, 0.0, 0.0, -0.0, -0.0, 74.0))


```

**Comment by fxcoudert** [comment]
```
elastic_doc = mpr.materials.elasticity.search(material_ids = ["mp-149"])
print(elastic_doc[0].elastic_tensor.ieee_format)
```

outputs

```
((153.0, 57.0, 57.0, -0.0, -0.0, 0.0), (57.0, 153.0, 57.0, -0.0, -0.0, 0.0), (57.0, 57.0, 153.0, -0.0, 0.0, 0.0), (-0.0, -0.0, -0.0, 74.0, 0.0, -0.0), (-0.0, -0.0, 0.0, 0.0, 74.0, -0.0), (0.0, 0.0, 0.0, -0.0, -0.0, 74.0))
```

which is rounded. I would like the tensor in the IEEE format, but not rounded. I don't want to reimplement it, you're right, I'd just like to know how to retrieve it with pymatgen, or transform the raw tensor (not rounded) again into the IEEE format.

**Comment by JaGeo** [comment]
@fxcoudert I think it is part of the `Tensor` class: https://pymatgen.org/pymatgen.core.html#pymatgen.core.tensors.Tensor.convert_to_ieee

In atomate2, the workflow for getting at it is maybe a bit clearer: https://materialsproject.github.io/atomate2/_modules/atomate2/common/schemas/elastic.html#ElasticDocument.from_stresses
It uses the `ElasticTensor` there which ultimately inherits from the `Tensor` above. Hard to see as it has two levels if inheritance.

I would assume a similar workflow has been used to generate the MP data.

**Comment by fxcoudert** [comment]
OK, so I think I can do:

```
from pymatgen.core.tensors import Tensor
Tensor.from_voigt(elastic_doc[0].elastic_tensor.raw).convert_to_ieee(elastic_doc[0].structure).voigt
```

It seems to return the same thing as `elastic_doc[0].elastic_tensor.ieee_format`, but not rounded. But maybe I'm just lucky because it's a cubic crystal anyway.

I still don't understand why the data is rounded in the MP database itself (rather than on display).

**Comment by JaGeo** [comment]
Tagging @tschaume as well as I don't find an immediate answer in emmet (https://github.com/materialsproject/emmet) which is the builder for the Materials Project data (I might have overlooked something or the data has been generated differently).

**Comment by matthewkuner** [comment]
@fxcoudert oops, I was testing on a large set of mp-ids. It's weird that mp-149 (which I just picked randomly to give you a single example (without checking, which was foolish)) is rounded, I agree...

**Reply by fxcoudert** [reply]
This code (linked below) seems to indicate the rounding is done on purpose: https://github.com/mjwen/emmet/blob/91ee6a40107571b423335e48e1769d1c44476104/emmet-core/emmet/core/elasticity.py#L283

I've tested many others, and they appear systematically rounded: mp-643989, mp-1345518, mp-22220, mp-1029903, etc.

**Comment by tschaume** [comment]
@JaGeo could [this line](https://github.com/materialsproject/emmet/blob/436566d834d359fc8e24b501ca74ce49163e885f/emmet-core/emmet/core/elasticity.py#L306) be the culprit?

**Reply by JaGeo** [reply]
Yep, looks like it. Did not find it while searching the code. Thanks! 

@fxcoudert at least we now see how the data is generated for the Materials Project. The why i cannot answer.

**Reply by tschaume** [reply]
Looks like it came in through https://github.com/materialsproject/emmet/pull/326 so @mjwen might have an explanation.

**Comment by mjwen** [comment]
Thanks all! 

Hi @fxcoudert, please see below. 

> 1. Why is mat.elastic_tensor.ieee_format rounded to 1 GPa for all entries?

The is a bit ad-hoc. All previous ieee elastic tensors were rounded to 1GPa (Although I am not 100% sure, I believe it is because we want not to make it very clumsy when displaying the values on the MP website). So, when I recalculate them, I simply adopted the previous convention. In retrospect, I should have kept more significant numbers. @tschaume might be able to fix this -- rerun parts of the builder. 

>  2. Where can I find the exact relationship between mat.elastic_tensor.ieee_format and mat.elastic_tensor.raw? What function is transforming one into the other? I would like to do it on my own, without rounding.

@matthewkuner pointed to the correct places,  just wanted to add one more point. ieee defines the "standard" orientation of each crystal based on the space group. The major difference between the ieee format and raw format is that the tensors are rotated. Of course, we also performed other symmetrization based on the space group as @matthewkuner mentioned. 

The solutions you provided above can get the unrounded one.

**Reply by tschaume** [reply]
Thanks @mjwen. Tagging @tsmathis who should be able to help with updating the builder to keep more significant digits the next time we run it.

**Reply by tsmathis** [reply]
Yep can do

**Comment by tschaume** [comment]
We were able to get some additional historical context from @kristinpersson about why `.round(0)` was chosen for the elastic tensor data: The benchmark at the time revealed about a 15% uncertainty resulting in the decision to report values in GPa without decimals. HTH.

**Reply by fxcoudert** [reply]
However, 15% uncertainty for smaller moduli is different than fixed rounding. Also, rounding can transform a valid (mechanically stable, definite positive matrix) tensor into a singular matrix.

If you ever revisit or recalculate these derived tensors, I would recommend avoiding rounding.

**Comment by hongyi-zhao** [comment]
Let me summarize the testing code for the issues discussed here:
```python
In [9]: from mp_api.client import MPRester
   ...: from pymatgen.core.tensors import Tensor
   ...: 
   ...: # Fetch elastic tensor data from Materials Project
   ...: with MPRester() as mpr:
   ...:     elastic_doc = mpr.materials.elasticity.search(material_ids=["mp-149"])
   ...: 
   ...:     if elastic_doc:  # Check if data exists
   ...:         # Get elastic tensor data in different formats
   ...:         raw_tensor = elastic_doc[0].elastic_tensor.raw
   ...:         ieee_tensor = elastic_doc[0].elastic_tensor.ieee_format
   ...: 
   ...:         # Convert to Pymatgen Tensor and transform to IEEE standard
   ...:         tensor = Tensor.from_voigt(raw_tensor)
   ...:         ieee_transformed = tensor.convert_to_ieee(elastic_doc[0].structure).voigt
   ...: 
   ...:         print("Raw elastic tensor:")
   ...:         print(raw_tensor)
   ...:         print("\nIEEE format (from API):")
   ...:         print(ieee_tensor)
   ...:         print("\nIEEE transformed tensor (using Pymatgen):")
   ...:         print(ieee_transformed)
   ...:     else:
   ...:         print("No elastic data found for mp-149")
   ...: 
Retrieving ElasticityDoc documents: 100%|████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 12052.60it/s]
Raw elastic tensor:
((153.19279267761982, 56.777185976088205, 56.777185976088205, -7.979727989493311e-17, -6.661338147750937e-16, 9.436895709313829e-16), (56.777185976088205, 153.19279267761982, 56.77718597608822, -6.661338147750937e-16, -7.979727989493311e-17, 6.839865079851815e-15), (56.777185976088205, 56.77718597608822, 153.19279267761985, -6.661338147750941e-16, 9.436895709313833e-16, -7.979727989493315e-17), (-7.979727989493311e-17, -6.661338147750937e-16, -6.661338147750941e-16, 74.25053622362908, 0.0, 1.2528910738651292e-32), (-6.661338147750937e-16, -7.979727989493311e-17, 9.436895709313833e-16, 0.0, 74.2505362236291, -2.0608663726559307e-15), (9.436895709313829e-16, 6.839865079851815e-15, -7.979727989493315e-17, 1.2528910738651292e-32, -2.0608663726559307e-15, 74.25053622362913))

IEEE format (from API):
((153.0, 57.0, 57.0, -0.0, -0.0, 0.0), (57.0, 153.0, 57.0, -0.0, -0.0, 0.0), (57.0, 57.0, 153.0, -0.0, 0.0, 0.0), (-0.0, -0.0, -0.0, 74.0, 0.0, -0.0), (-0.0, -0.0, 0.0, 0.0, 74.0, -0.0), (0.0, 0.0, 0.0, -0.0, -0.0, 74.0))

IEEE transformed tensor (using Pymatgen):
[[ 1.53192793e+02  5.67771860e+01  5.67771860e+01 -1.71248977e-15
  -7.85490033e-16  5.40602707e-15]
 [ 5.67771860e+01  1.53192793e+02  5.67771860e+01 -5.40602707e-15
  -1.00148357e-31  5.40602707e-15]
 [ 5.67771860e+01  5.67771860e+01  1.53192793e+02 -5.40602707e-15
   7.85490033e-16  1.71248977e-15]
 [-1.71248977e-15 -5.40602707e-15 -5.40602707e-15  7.42505362e+01
   2.23951367e-15 -3.54252944e-32]
 [-7.85490033e-16 -5.75211077e-32  7.85490033e-16  2.23951367e-15
   7.42505362e+01 -2.23951367e-15]
 [ 5.40602707e-15  5.40602707e-15  1.71248977e-15 -5.10157680e-32
  -2.23951367e-15  7.42505362e+01]]

```
